[16.0][MIG] account_product_fiscal_classification (from 14.0 to 16.0)

legalsylvain opened the pull request:
Migration of ``account_product_fiscal_classification`` from 14.0 to 16.0.
- remove ``account_product_fiscal_classification_test``. (now all tests are done in the same module).
- increase coverage of the module. 81,9% -> 86,5%
- code simplification. (python lines : 526 -> 444) = -15%
- Remove ``usage_group_id`` field on classification and fiscal classification on product.category, replace by generic new model ``account.product.fiscal.rule``.
- Improve installation of the module. (now install module on huge database takes second, before minutes).
- remove double write access on DB when creating or writing ``product.template`` models.
- improve and update documentation.

### Description

This module extends the Odoo Account module to introduce Fiscal Classifications.

A fiscal classification is an association of purchase and sale taxes that have to be set on your products.

![product_template_form](https://user-images.githubusercontent.com/3407482/196738348-1329ba22-441a-4207-a73d-2da0885b8df3.png)

This new design provides the following improvments.

* It is more simple (less clicks quantity) to configure correctly a product. (You only have to set a fiscal classification, instead of setting one or many sale taxes, then click on purchase tab and set one or many purchase taxes.)

* Prevent users to select incompatible purchase and sale taxes. French Exemple: A product can not be configured with:

  * Purchase Taxes: 5.5%
  * Sale Taxes: 20%

* Optionaly, you can restrict the usage of some fiscal classifications to some product categories.
  French Exemple : A product set in the category "Beer and wine" can not have a sale tax 5,5%. (20% is mandatory)


### Configure

* Go to "Accounting > Configuration > Accounting > Fiscal Classifications".

* Create new classification, with correct purchase and sale taxes.

![fiscal_classification_form](https://user-images.githubusercontent.com/3407482/196735119-a72de8d0-fb96-409d-ab67-8e2ca84e6afe.png)

**Note**

When installing the module, it will create fiscal classifications for all the taxes combinations that exist in your database.
If some are bad, you can massively move the products for a bad classification to the good one, using the button "Move products into another fiscal classification".

**Optional settings**

Once the fiscal classifications are created, you can create rules to avoid users to use some classifications that don't make sense for some categories.

* Go to "Accounting > Configuration > Accounting > Fiscal Classification Rules"
* Create new rules.

![fiscal_classification_rule__wine](https://user-images.githubusercontent.com/3407482/196735160-db14146c-38e3-40d0-a21f-204f23a34ff9.png)


Every time a new product is created, the rules that matches the product setting will be checked **in the displayed order**.

* If one rule is set to 'Allow', the next rules are ignored.
* If one rule is set to 'Forbid', an error is raised.

![fiscal_classification_rule__wine_warning](https://user-images.githubusercontent.com/3407482/196735185-29d79c5c-3a48-4327-8689-0483d7c686e5.png)

rvalyi committed:
[MERGE] merged useful tax logic modules from 5.0 branch (they are required for the Brazilian localization)

rvalyi committed:
[FIX] account_product_fiscal_classification: temporary workaround to pass the ir.rule.group refactoring

rvalyi committed:
[FIX] account_product_fiscal_classification: fixed the way user avoid messing with product taxes from an other (sub) company

Commit:
changed fiscal_classification_id_change method in product.product and added company_id.id parameter in search method in account.tax objects

Commit:
migrate module account_product_fiscal_classification to compatible with OpenERP v6

Commit:
added folder security with access rules for objects in modules account_product_fiscal_classification and account_fiscal_position_rule and remove old files __terp__.py

rvalyi committed:
[FIX] account_product_fiscal_classification: typically, service products do not require any fiscal classification

renatonlima committed:
account_fiscal_position_rule, account_fiscal_position_rule_sale refactoring and changed in account_product_fiscal_classification module product.product to product.template

renatonlima committed:
account_fiscal_position_rule modules and account_product_fiscal_classification changed header files and included in __openerp__.py explicit AGPL-3 license

renatonlima committed:
account_product_fiscal_classification, add Update All Produtos Button in Fiscal Classification

renatonlima committed:
[FIX] onchange methods in account_fiscal_position_rule_sale and account_product_fiscal_classification

renatonlima committed:
updated wizards from account_fiscal_position_rule and

Commit:
Fixed account.invoice views from account_fiscal_position_rule module

Commit:
clean up account_product_fiscal_classification

sebastienbeau committed:
[REF] Clean syntax

sebastienbeau committed:
[REF] Update copyright

rvalyi committed:
[MIGR] fixed tree view

rvalyi committed:
[MIGR] view fix for v7

renatonlima committed:
Improved code format, changed views for migration to OpenERP 7.0 and methods in in object account.product.fiscal.classification

renatonlima committed:
[MIGR] account_product_fiscal_classification module migration to OpenERP v7

renatonlima committed:
[FIX] refactored onchange method and field definition

Commit:
product account tab view correction, for better reading

gurneyalex committed:
mark all addons as unported

legalsylvain committed:
[REF] rename module 1/2;

legalsylvain committed:
[REF] renaming model;

legalsylvain committed:
[REF] change function name according to the new objective of the function;

legalsylvain committed:
[ADD] add demo data and change rules to manage and test multi company context;

legalsylvain committed:
[REF]

legalsylvain committed:
[REF] move init function into 'post_init_hook' part. courtesy : @StefanRijnhart;

legalsylvain committed:
[REF] use new API; Remove useless code, as the init function is now only called at the installation, and no during update anymore;

legalsylvain committed:
[FIX] bad path for screenshots;

legalsylvain committed:
[FIX] flake8;

legalsylvain committed:
[REF] improve comments about how classification_id field is set as required;

legalsylvain committed:
[ADD] pot file ; fr translation file ; set name field as translatable;

legalsylvain committed:
[MRG] add pt_BR.po file;

legalsylvain committed:
[ADD] code and description on account_product_fiscal_classification model;j

oca-transbot committed:
OCA Transbot updated translations from Transifex

renatonlima committed:
[FIX] invalid name fields taxes

renatonlima committed:
added fiscal classification template model

renatonlima committed:
added action, menu and views to fiscal classification model

renatonlima committed:
added wizard fiscal classification model

renatonlima committed:
added action, menu and views to fiscal classification wizard

added missing file from last commit

renatonlima committed:
added tests

renatonlima committed:
pep8

pep8

[REF] One model per file

renatonlima committed:
change version

renatonlima committed:
rename folders according OCA Guidelines

change version in README.md

pep8

rvalyi committed:
fixed wrong xml_id ref -> tests 1,2,3 and 4 pass. 5 and 6 fail.

mileo committed:
[FIX] Revert changing in user group and fix remaining tests

mileo committed:
[FIX] Pep8

[FIX Flake8

mileo committed:
magically turn testing on by renaming tests/tests.py to tests/test_classification.py; tests are broken

Conflicts:
	account_product_fiscal_classification/tests/__init__.py

rvalyi committed:
[MIG] Make modules uninstallable

mbcosta committed:
Migration the views, menus, add depends and make module installable.

mbcosta committed:
Migration demo datas an add license header.

mbcosta committed:
[MIG] Add attribute name to allow inherit the view.

oca-transbot committed:
OCA Transbot updated translations from Transifex

legalsylvain committed:
[REF] OCA convention; (light header);

legalsylvain committed:
[ADD] classification restriction by product category;

legalsylvain committed:
[FIX] set api.multi in a function; (bug if we changed classification for more than one product;)

legalsylvain committed:
[IMP] improve display;

legalsylvain committed:
[FIX] bug : it was not possible to delete a classification after mass change;

sbidoul committed:
[FIX] remove en.po that was erroneously created by transbot

oca-transbot committed:
OCA Transbot updated translations from Transifex

mbcosta committed:
Change field Fiscal Position to be required. (#53)

[IMP] account_fiscal_position_rule: Change field Fiscal Position to be required

oca-transbot committed:
OCA Transbot updated translations from Transifex

legalsylvain committed:
[FIX] incorrect order when loading demo data

legalsylvain committed:
[ADD] author

legalsylvain committed:
[FIX] remove invalid leaf in V9.0 on account.tax / account.tax.template parent_id fields

oca-transbot committed:
OCA Transbot updated translations from Transifex

rvalyi committed:
[MIG] Make modules uninstallable

rvalyi committed:
[MIG] Rename manifest files

renatonlima committed:
[MIGR] fix header files

renatonlima committed:
[MIGR] update module metadata

renatonlima committed:
[MIGR] fix import

renatonlima committed:
[MIGR] product.template object

renatonlima committed:
[MIGR] product.category object

renatonlima committed:
[MIGR] account.product.fiscal.classification.model object

renatonlima committed:
[MIGR] account.product.fiscal.classification object

renatonlima committed:
[MIGR] security xml file

renatonlima committed:
[MIGR] format xml files

renatonlima committed:
[MIGR] migrate yml files

renatonlima committed:
[MIGR] fix file headers

renatonlima committed:
added security rule to fiscal classification template

bealdav committed:
IMP category views

bealdav committed:
FIX conflict revert

legalsylvain committed:
[FIX] remove useless tab moves, reducing diff during review

oca-transbot committed:
OCA Transbot updated translations from Transifex

sbidoul committed:
remove obsolete .pot files [ci skip]

oca-travis committed:
[UPD] Update account_product_fiscal_classification.pot

pedrobaeza committed:
[FIX] account_product_fiscal_classification: RST syntax

Commit:
[MIG] 12.0

legalsylvain committed:
[MIG] finish to port account_product_fiscal_classification in V12
[REF] split tests into the two modules:
- account_product_fiscal_classification has all the tests that doesn't require l10n_generic_coa;
- account_product_fiscal_classification_test has all the tests that requires l10n_generic_coa;
[REF] Improve the generation of the classifications based on the classification templates
- link fiscal_classification.template to account.chart.template
- when loading a chart template, all the fiscal positions are created,
[IMP] Menus. Display classification and templates in according V12 sub menu.
[REF] Demo data :
- remove useless demo data ;
- move demo data into the active module. (except demo data that depends on l10n_generic_coa)
[REF] OCA convention
- create wizard folder;
- create one xml view per model;
- create readme folder for account_product_fiscal_classification_test module;
[FIX] restore post_init hook function that was dropped in the migration to V12.
[REF] remove useless abstract model account.product.fiscal.classification.model
[ADD] add possibility to restrict usage of fiscal classification to a given group of users.
[ADD] dependencies to account-financial-tools
[REF] update fr translation

oca-travis committed:
[UPD] Update account_product_fiscal_classification.pot

OCA-git-bot committed:
[ADD] icon.png

OCA-git-bot committed:
[UPD] README.rst

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-fiscal-rule-12.0/account-fiscal-rule-12.0-account_product_fiscal_classification
Translate-URL: https://translation.odoo-community.org/projects/account-fiscal-rule-12-0/account-fiscal-rule-12-0-account_product_fiscal_classification/

alvarorib committed:
Translated using Weblate (Portuguese)

Currently translated at 56.5% (35 of 62 strings)

Translation: account-fiscal-rule-12.0/account-fiscal-rule-12.0-account_product_fiscal_classification
Translate-URL: https://translation.odoo-community.org/projects/account-fiscal-rule-12-0/account-fiscal-rule-12-0-account_product_fiscal_classification/pt/

bealdav committed:
FIX prd_classif: set taxes on all companies sharing classif

oca-travis committed:
[UPD] Update account_product_fiscal_classification.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
account_product_fiscal_classification 12.0.1.1.0

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-fiscal-rule-12.0/account-fiscal-rule-12.0-account_product_fiscal_classification
Translate-URL: https://translation.odoo-community.org/projects/account-fiscal-rule-12-0/account-fiscal-rule-12-0-account_product_fiscal_classification/

dreispt committed:
[FIX] account_product_fiscal_classification: fix duplicate XMLId

oca-travis committed:
[UPD] Update account_product_fiscal_classification.pot

OCA-git-bot committed:
account_product_fiscal_classification 12.0.1.2.0

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-fiscal-rule-12.0/account-fiscal-rule-12.0-account_product_fiscal_classification
Translate-URL: https://translation.odoo-community.org/projects/account-fiscal-rule-12-0/account-fiscal-rule-12-0-account_product_fiscal_classification/

sysadminmatmoz committed:
Translated using Weblate (Slovenian)

Currently translated at 100.0% (66 of 66 strings)

Translation: account-fiscal-rule-12.0/account-fiscal-rule-12.0-account_product_fiscal_classification
Translate-URL: https://translation.odoo-community.org/projects/account-fiscal-rule-12-0/account-fiscal-rule-12-0-account_product_fiscal_classification/sl/

legalsylvain committed:
[FIX] allow to see products related to fiscal classification

OCA-git-bot committed:
account_product_fiscal_classification 12.0.1.3.0

bealdav committed:
IMP prd_fisc_classif: track change on product model

oca-travis committed:
[UPD] Update account_product_fiscal_classification.pot

OCA-git-bot committed:
[UPD] README.rst

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-fiscal-rule-12.0/account-fiscal-rule-12.0-account_product_fiscal_classification
Translate-URL: https://translation.odoo-community.org/projects/account-fiscal-rule-12-0/account-fiscal-rule-12-0-account_product_fiscal_classification/

Commit:
[REF] account_product_fiscal_classification: Black python code

Commit:
[MIG] account_product_fiscal_classification: Migration to 13.0

Commit:
[IMP] account_product_fiscal_classification, account_product_fiscal_classification_test: black, isort, prettier

Commit:
[MIG] account_product_fiscal_classification: Migration to 14.0

oca-travis committed:
[UPD] Update account_product_fiscal_classification.pot

OCA-git-bot committed:
[UPD] README.rst

legalsylvain committed:
[REF] adopting account_product_fiscal_classification and account_product_fiscal_classification_test

OCA-git-bot committed:
[UPD] README.rst

dreispt committed:
[REF] account_product_fiscal_classification: pre-commit run fixes

OCA-git-bot committed:
account_product_fiscal_classification 14.0.1.0.1

legalsylvain committed:
[IMP] account_product_fiscal_classification: black, isort, prettier

legalsylvain committed:
[PORT] account_product_fiscal_classification: port module from 14.0 to 16.0
- remove ``account_product_fiscal_classification_test``. (now all tests are done in the same module).
- increase coverage of the module.
- code simplification.
- Remove ``usage_group_id`` field on classification and fiscal classification on product.category, replace by generic new model ``account.product.fiscal.rule``.
- Improve installation of the module. (now install module on huge database takes second, before minutes).
- remove double write access on DB when creating or writing ``product.template`` models.
- improve and update documentation.

legalsylvain committed:
[UPD] translation

alexis-via commented:
Great to see the end of the double write... i guess we can start to use this module now :)

I would suggest the use of "_check_company_auto = True" on account.product.fiscal.classification. I'll make a PR.

legalsylvain commented:
> Great to see the end of the double write... i guess we can start to use this module now :)

Nice ! 

> I would suggest the use of "_check_company_auto = True" on account.product.fiscal.classification. I'll make a PR.

you can do it against legalsylvain:16.0-mig-account_product_fiscal_classification, I'll merge it quickly.

alexis-via commented:
If we use _check_company_auto = True, it would be more appropriate on product.template, not on account.product.fiscal.classification. There is no _check_company_auto by default on product.template, so...
BTW, I think we should align the behavior of the default value for company_id of account.product.fiscal.classification with the behavior on product.template. It seems to me that, by default, company_id is False on product.template (it is a configuration param somewhere ?) ; so maybe should also have company_id = False by default on account.product.fiscal.classification, don't you think ?

rafaelbn commented:
Thank you!

@moduon MT-1618

rafaelbn commented on account_product_fiscal_classification/__manifest__.py:
```diff
@@ -0,0 +1,39 @@
+    "maintainers": ["legalsylvain"],
+    "website": "https://github.com/OCA/account-fiscal-rule",
+    "license": "AGPL-3",
+    "depends": ["account", "account_usability"],
```
Is it required to depends in `account_usability`? I understand `account_usability `but I don't use this module and I would like to maintain without using it.

Just a question. Thank you! 😄

legalsylvain replied:
Hi @rafaelbn. account_usability module (previously named account_menu module in v<16) is a simple module that adds menu entries. It is required to put the menu of the classification template in the correct place. IMO : 

- the module is well maintained. (present in all the version)
- it is easy (mostly xml)

rvalyi replied:
> Hi @rafaelbn. account_usability module (previously named account_menu module in v<16) is a simple module that adds menu entries. It is required to put the menu of the classification template in the correct place. IMO : 
> 
> - the module is well maintained. (present in all the version)
> - it is easy (mostly xml)

Hello, Sylvain, I have the same doubt for other modules I want to port to v16. One of the doubts I have is if people will complain if it doesn't play well with Odoo EE. Do you know if account_menu works decently along with Odoo EE?

legalsylvain replied:
Hi. 

> Do you know if account_menu works decently along with Odoo EE?

I don't understand why you ask me ? ;-) 
I don't have access to the EE enterprise code, and your akrétion team has access.
So only you can say if there are incompability ! 

But technically, we have to understand that OCA modules are only "guaranteed" for the CE version, by design (there is no OCA-EE runboat, and no OCA-EE test).

rvalyi replied:
yes, I understand. That being said lot's of OCA modules got little modifications to play it nice with Odoo EE as many OCA users also have customers with EE. As for me I don't have access to a v16 EE repo and that's not a big deal/blocker for me, just "a better have" in case you knew the answer about compatibility.

rafaelbn reviewed (commented):
We have the need to set fiscal positions to product that over write the fiscal position of the partner

- Product fiscal position has more priority than partner fiscal position

But not always, only on some specific products

rafaelbn commented:
/ocabot migration account_product_fiscal_classification

legalsylvain commented:
Hi @rafaelbn 

> We have the need to set fiscal positions to product that over write the fiscal position of the partner

>    Product fiscal position has more priority than partner fiscal position

> But not always, only on some specific products

* I'm not sure to understand your need, could you elaborate with a real exemple
* At the first sight, adding fiscal position on product should be done in another module. The objective of this one is only to improve fiscal taxes settings on product.

@alexis-via 

> It seems to me that, by default, company_id is False on product.template (it is a configuration param somewhere ?) ; so maybe should also have company_id = False by default on account.product.fiscal.classification, don't you think ?

hum... 
- in V12 (and before) : default company of product template is the current one. ([See](https://github.com/odoo/odoo/blob/12.0/addons/product/models/product_template.py#L110))
- in V13 (and after) : company of product template is False by default. ([See](https://github.com/odoo/odoo/blob/13.0/addons/product/models/product_template.py#L121))

I don't like the new design, because that's not the one I implemented in my multicompany context !
But well, you're probably right, we should have a consistent logic.
I removed the default on account.product.fiscal.classification.

@rafaelbn, @alexis-via : could you update your review ? 

thanks, and happy new years !

Shide reviewed (approved):
LGTM

rafaelbn commented:
/ocabot merge partch

rafaelbn commented:
/ocabot merge patch

rafaelbn commented:
/ocabot merge patch

rafaelbn commented:
@legalsylvain could you check failure please? Thank you!
